src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1588 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_INTRUSIVELIST_HPP
+#define SHARE_UTILITIES_INTRUSIVELIST_HPP
+
+#include "memory/allStatic.hpp"
+#include "metaprogramming/enableIf.hpp"
+#include "metaprogramming/logical.hpp"
+#include "utilities/align.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/macros.hpp"
+#include <type_traits>
+
+class IntrusiveListEntry;
+class IntrusiveListImpl;
+
+/**
+ * The type of a function for accessing the list entry of an IntrusiveList's
+ * element type T.  Such a function takes a reference to const T and returns a
+ * reference to const IntrusiveListEntry.
+ */
+template<typename T>
+using IntrusiveListEntryAccessor =
+  const IntrusiveListEntry& (*)(std::add_const_t<T>&);
+
+/**
+ * The IntrusiveList class template provides a doubly-linked list in
+ * which the links between elements are embedded directly into objects
+ * contained in the list.  As a result, there are no copies involved
+ * when inserting objects into the list or referencing list objects,
+ * and removing an object from a list need not involve destroying the
+ * object.
+ *
+ * To be used in a IntrusiveList, an object must have a
+ * IntrusiveListEntry member.  An IntrusiveList is associated with the
+ * class of its elements and the entry member.
+ *
+ * An object can be in multiple lists at the same time, so long as
+ * each list uses a different entry member.  That is, the class of the
+ * object must have multiple IntrusiveListEntry members, one for each
+ * list the object is simultaneously an element.
+ *
+ * The API for IntrusiveList is modelled on the containers provided by
+ * the C++ standard library.  In particular, iteration over the
+ * elements is provided via iterator classes.
+ *
+ * IntrusiveLists support polymorphic elements.  Because the objects
+ * in a list are externally managed, rather than being embedded values
+ * in the list, the actual type of such objects may be more specific
+ * than the list's element type.
+ *
+ * * T is the class of the elements in the list.  Must be a class type,
+ * possibly const-qualified.
+ *
+ * * get_entry is a function of type IntrusiveListEntryAccessor<T> used
+ * for accessing the IntrusiveListEntry subobject of T used by this list.
+ *
+ * * has_size determines whether the list has a size()
+ * operation, returning the number of elements in the list.  If the
+ * operation is present, it has constant-time complexity.  The default
+ * is to not provide a constant-time size() operation.
+ *
+ * * Base is the base class for the list.  This is typically
+ * used to specify the allocation class, such as CHeapObj<>.  The default
+ * is void, indicating the list is not derived from an allocation class.
```

Comment:
I'm not certain this Base class for allocation support is actually needed.  I remember one of the alternatives had
(or used to have?) allocation base class support, but haven't found it when I looked recently.  But we have a lot
of these doubly-linked-lists in HotSpot.  Do we have a use-case for a "heap" allocated bare (as in not
embedded in some other object) list?  Removing it would save ~25 lines of code/comments.

Reply:
Did a quick skim as well and yes, quite a few ad-hoc lists in the code base. Did not see a case needing this and I think removing this makes sense if we don't see an obvious use-case.

Reply:
OK, removed.  We can always add it back if we find a use-case.